src/gensearch.jl

```diff
@@ -0,0 +1,372 @@
+
+Follows the algorithm given in  https://arxiv.org/abs/2211.16468. 
+"""
+function find_min_frontdoor_adjustment(g, X, Y, I = Set{eltype(g)}(), R = setdiff(Set(vertices(g)), X, Y))
+    X, Y, I, R = toset.((X, Y, I, R))
+    # write explanation for the pass rules
+    Zii = find_frontdoor_adjustment(g, X, Y, I, R)
+    Zii == false && return false 
+    function Za_pass(pe, ne, v, w)
+        pe in [INIT, LEFT] && ne == LEFT && !(v in Zii) && !inunion(w, X, Y) && return true
+        return false
+    end
+    Za = intersect(gensearch(g, Y, Za_pass), Zii) 
+    function Zxy_pass(pe, ne, v, w)
+        pe in [INIT, RIGHT] && ne == RIGHT && !(v in Za) && !inunion(w, X, Y, I) && return true
+        return false
+    end
+    Zxy = intersect(gensearch(g, X, Zxy_pass), Za)
+    function Zzy_pass(pe, ne, v, w)
+        pe in [INIT, LEFT] && ne == LEFT && !inunion(w, X, I, Zxy) && return true
+        pe in [LEFT, RIGHT] && ne == RIGHT && !(w in X) && !inunion(v, I, Za) && return true
+        pe == RIGHT && ne == LEFT && inunion(v, I, Za) && !inunion(w, X, I, Zxy) && return true
+        return false
+    end
+    Zzy = intersect(gensearch(g, union(I, Zxy), Zzy_pass), Za) 
+    return union(I, Zxy, Zzy)
+end
+
+struct ConstraintIterator{T<:Integer, F<:Function}
+    g::SimpleDiGraph{T}
+    X::Set{T}
+    Y::Set{T}
+    I::Set{T}
+    R::Set{T}
+    find::F
+end
+
+function downwards(state, I, R)
+    v = first(setdiff(R, I))
+    push!(state, (:up, :I, v, I, R))
+    push!(state, (:down, :I, v, I, R))
+    push!(state, (:up, :R, v, I, R))
+    push!(state, (:down, :R, v, I, R))
+end
+
+# TODO: do this more elegantly
+function Base.iterate(A::ConstraintIterator)
+    R = deepcopy(A.R)
+    I = deepcopy(A.I)
+    state = Vector{Tuple{Symbol, Symbol, Int64, Set{Int64}, Set{Int64}}}()
+    A.find(A.g, A.X, A.Y, I, R) == false && return nothing
+    issetequal(I, R) && return I, state
+    downwards(state, I, R)
+    Base.iterate(A, state)
+end
+
+function Base.iterate(A::ConstraintIterator, state)
+    while !isempty(state)
+        dir, set, v, I, R = pop!(state)
+        dir == :down && set == :I && push!(I, v)
+        dir == :up && set == :I && delete!(I, v)
+        dir == :down && set == :R && delete!(R, v)
+        dir == :up && set == :R && push!(R, v)
+        A.find(A.g,  A.X, A.Y, I, R) == false && continue
+        issetequal(I, R) && return deepcopy(I), state
+        if dir == :down
+            downwards(state, I, R)
+        end
+    end
+    return nothing
+end
+
+Base.IteratorSize(::ConstraintIterator) = Base.SizeUnknown()
+
+"""
+    list_dseps(g, X, Y, I = Set{eltype(g)}(), R = setdiff(Set(vertices(g)), X, Y))
+
+List all d-separators `Z` with `I subseteq Z subseteq R` for sets of vertices `X` and `Y` in `g`. 
+"""
+function list_dseps(g, X, Y, I = Set{eltype(g)}(), R = setdiff(Set(vertices(g)), X, Y))
+    X, Y, I, R = toset.((X, Y, I, R))
```

Comment:
E.g. the cast on X and Y is not necessary?

Reply:
You mean for list functions specifically or also in other places?

Here ```find_dsep``` would cast X at every call (which happens during ```list_dseps``) if we don't do the cast here. It's true that ```find_dsep``` also doesn't really need X to be a Set, however, e.g. ```find_covariate_adjustment``` and other functions do (because membership v in X is tested repeatedly), so at that point I thought it was cleanest to cast everything.